File: mscg/timer.py
from time import time

class TIMER:
    start = 0.0
    last = 0.0
    recs = {}
    
    @staticmethod
    def reset():
        TIMER.start = time()
        TIMER.last = time()
        TIMER.recs = {}
    
    @staticmethod
    def click(name, f=None):
        now = time()
        
        if name is not None:
            if name not in TIMER.recs:
                TIMER.recs[name] = [0, 0.0]
            
            TIMER.recs[name][0] += 1
            TIMER.recs[name][1] += now - TIMER.last
        
        TIMER.last = now
    
    @staticmethod
    def report(screen = True):
        now = time()
        total = now - TIMER.start
        timed = 0.0
        
        w = 42
        z = []
        
        z.append("=" * w)
        z.append("            Timing Statistics")
        z.append("-" * w)
        z.append(" %-10s %7s %10s %10s" % ("module", "count", "elapsed", "fraction"))
        z.append("-" * w)
        
        for name in TIMER.recs:
            rec = TIMER.recs[name]
            timed += rec[1]
            z.append(" %-10s %7d %10.2f %9.1f%%" % (name[:10], rec[0], rec[1], rec[1]/total * 100))
        
        z.append("-" * w)
        other = total - timed
        z.append(" %-10s %7s %10.2f %9.1f%%" % ("other", "", other, other/total * 100))       
        z.append(" %-10s %7s %10.2f %9.1f%%" % ("total", "", total, 100.0))
        z.append("=" * w)
        
        if screen:
            print("\n".join(z))
        
        return z
        

File: mscg/core/src/timer.cpp
#include "timer.h"
#include "std.h"

const char *timer_names[] = {
    "IO", "LIST", "TABLE", "MATRIX", "SOLVER",
    0
};

Timer::Timer()
{
    ntimers = 0;
    while(timer_names[ntimers]) ntimers++;
    
    count = new int[ntimers];
    seconds = new double[ntimers];
    
    for(int i=0; i<ntimers; i++)
    {
        count[i] = 0;
        seconds[i] = 0.0;
    }
}

Timer::~Timer()
{
    delete [] count;
    delete [] seconds;
}

void Timer::reset()
{
    last = clock();
    start = last;
}

void Timer::click(int id)
{
    clock_t now = clock();
    
    count[id]++;
    seconds[id] += (double)(now - last)/CLOCKS_PER_SEC;
    
    last = now;
}

void Timer::report()
{
    clock_t now = clock();
    
    printf("Total time: %0.2lf seconds\n", (double)(now - start)/CLOCKS_PER_SEC);
    
    for(int i=0; i<ntimers; i++)
        printf("%8s %6.2lf (%d)\n", timer_names[i], seconds[i], count[i]);
}